src/libmongoc/src/mongoc/mongoc-server-monitor.c

```diff
@@ -0,0 +1,1329 @@
+      GOTO (fail);
+   }
+
+   _mongoc_rpc_swab_from_le (&rpc);
+   memcpy (&msg_len, rpc.msg.sections[0].payload.bson_document, 4);
+   msg_len = BSON_UINT32_FROM_LE (msg_len);
+   if (!bson_init_static (
+          &reply_local, rpc.msg.sections[0].payload.bson_document, msg_len)) {
+      bson_set_error (error,
+                      MONGOC_ERROR_PROTOCOL,
+                      MONGOC_ERROR_PROTOCOL_INVALID_REPLY,
+                      "Malformed BSON payload from server");
+      GOTO (fail);
+   }
+
+   bson_copy_to (&reply_local, ismaster_reply);
+   server_monitor->more_to_come =
+      (rpc.msg.flags & MONGOC_MSG_MORE_TO_COME) != 0;
+
+   ret = true;
+fail:
+   if (!ret) {
+      bson_init (ismaster_reply);
+   }
+   _mongoc_buffer_destroy (&buffer);
+   return ret;
+}
+
+/* Send and receive an awaitable ismaster.
+ *
+ * Called only from server monitor thread.
+ * Locks server monitor mutex.
+ * May block for up to heartbeatFrequencyMS waiting for reply.
+ */
+static bool
+_server_monitor_awaitable_ismaster (mongoc_server_monitor_t *server_monitor,
+                                    const bson_t *topology_version,
+                                    bson_t *ismaster_reply,
+                                    bool *cancelled,
+                                    bson_error_t *error)
+{
+   bson_t cmd = BSON_INITIALIZER;
+   bool ret = false;
+
+   bson_append_int32 (&cmd, "isMaster", 8, 1);
+   _server_monitor_append_cluster_time (server_monitor, &cmd);
+   bson_append_document (&cmd, "topologyVersion", 15, topology_version);
+   bson_append_int32 (
+      &cmd, "maxAwaitTimeMS", 14, server_monitor->heartbeat_frequency_ms);
+   bson_append_utf8 (&cmd, "$db", 3, "admin", 5);
+
+   if (!_server_monitor_awaitable_ismaster_send (server_monitor, &cmd, error)) {
+      GOTO (fail);
+   }
+
+   if (!_server_monitor_awaitable_ismaster_recv (
+          server_monitor, ismaster_reply, cancelled, error)) {
+      bson_destroy (ismaster_reply);
+      GOTO (fail);
+   }
+
+   ret = true;
+fail:
+   if (!ret) {
+      bson_init (ismaster_reply);
+   }
+   bson_destroy (&cmd);
+   return ret;
+}
+
+/* Attempt to enable server monitor for awaitable ismaster.
+ *
+ * Called only from server monitor thread.
+ * Caller must not hold locks.
+ * Locks server monitor mutex.
+ * Returns true if awaitable was enabled (or is already enabled).
+ * Returns false if awaitable cannot be enabled.
+ */
+static bool
+_server_monitor_enable_awaitable_ismaster (
+   mongoc_server_monitor_t *server_monitor)
+{
+   if (server_monitor->initiator) {
+      /* Disable for custom initiator. */
+      MONITOR_LOG_WARNING (server_monitor, "unable to enable server monitor "
+                                           "streaming due to custom stream "
+                                           "initiators");
+      return false;
+   }
+   bson_mutex_lock (&server_monitor->shared.mutex);
+   if (server_monitor->shared.interrupt) {
```

Comment:
TODO: let's see if we can remove this interrupt.
Poll with a short timeout and check if cancelled instead.